CBPs/cbp-0002.md

```diff
@@ -0,0 +1,632 @@
+
+7. approve/transferFrom (Approve is a pairing function for TransferFrom)
+
+- Question: There was a disagreement about whether the approval/transfer should be removed in the discussion on the forum
+
+- Consideration:
+
+  approve/transferFrom appears in ERC20 mainly because:
+
+  > Using Ethereum's native ETH token, you can call the smart contract function and send ETH to the contract at the same time. This is done using payable. But because the ERC20 token itself is a smart contract, it is not possible to directly send the token to the smart contract while calling one of its functions; therefore, the ERC20 standard allows smart contracts to transfer tokens on behalf of the user-using the transferFrom() function. For this, users need to allow smart contracts to transfer these tokens on their behalf
+
+  However, in the Dex and lending scenarios of Ethereum, Approve is often accompanied by the possibility of simultaneous operation of two tokens. Approve can avoid the repeated payment problem which transaction brought about, has a good supplementary use scenario for transfer.
+
+- Solution: Approve/transferFrom should be retained
+
+8. TransferAndCall vs Receiver Notify
+- Question: which option is more suitable
+- Consideration:
+  - Notify can meet the basic notification needs. Although it cannot support better flexibility, it is sufficient to meet the transfer scenario
+  - TransferAndCall provides better flexibility, but it depends on the transfer caller to fully understand the method and parameters corresponding to the call, which is not needed for most transfer scenarios
+  - For security reasons, do not call the canister of the location inside the canister,[why?Inter-canister calls](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister)
+  
+- Solution: 
+  - Neither is supported
+
+9. ApproveAndCall VS TransferAndCall
+
+- Question: We compare ApproveAndCall and TransferAndCall. ApproveAndCall and TransferAndCall are two sets of non-atomic operations, there is no difference essentially. Which one should be retained?
+
+- Consideration: In some scenarios, when multiple Tokens need to be transferred at the same time, TransferAndCall can not meet such needs. After approval, execute transferFrom in the final call to pay multiple tokens at once
+- For security reasons like Q8
+
+- Solution: call is not supported
+
+10. Nonce
+
+- Question: What Is Nonce In Ethereum? every transaction has a nonce. The nonce is the number of transactions sent from a given address. Each time you send a transaction, the nonce increases by 1. There are rules about what transactions are valid, and the nonce is used to enforce some of these rules. Does Dfinity's token standard need to incorporate a nonce mechanism?
+
+- Consideration:If the nonce mechanism is supported, it can prevent users from repeatedly submitting transactions
```

Comment:
The IC already prevents the same message being sent twice, using a nonce in the message.
This nonce is there to allow stable transaction IDs to be generated offline for the benefit of exchanges.

If you want these tokens to be one day tradable on an exchange you probably want to keep the nonce.

Reply:
IC's nonce does not give developers a way to use the nonce in order for developers to avoid sending the same transaction, which is important. Of course, if IC had a clear way to support developers using nonce, there would be no need for nonce to be in the token standard

Reply:
Ah sorry I should be more clear. 
Every ingress message sent to the IC has a [nonce generated](https://github.com/dfinity/interface-spec/blob/public-antora/modules/interface-spec/pages/index.adoc#request-call).
In addition there is a second nonce which is included in the block.
The exchanges enforce an invariant that each transaction must have a statically derivable globally unique request ID. This second nonce helps ensure that each transaction is globally unique.
This is also why we have the created_at field in ICP.

Reply:
Thank you very much for your patience, I think this approach makes sense, using create_at as the second nonce would be simpler and I believe developers would like this approach more. I will adjust the proposal according to your suggestion